proposals/text/3784-input-output-shaper.md

```diff
@@ -0,0 +1,247 @@
+
+```yaml
+            components:
+            - name: shaper
+              params:
+                inputs:
+                    query:
+                      func: expand
+                      output: questions
+                      params:
+                        expand_target: query
+                        size:
+                          func: len
+                          params:
+                            - documents
+                    documents:
+                      func: contract_docs
+                      output: documents
+                      params:
+                        docs: documents
+                        delimiter: " "
+                        num_tokens:
+                            func: len
+                            params:
+                                - questions
+              type: InputOutputShaper
+            pipelines:
+              - name: query
+                nodes:
+                  - name: shaper
+                    inputs:
+                      - Query
+```
+
+### Security
+
+The InputOutputShaper component prevents arbitrary code execution. Only the functions we have specified in the
+InputOutputShaper component are allowed to be invoked. The functions are specified in the constructor using immutable
+data structures.
+
+This security measure might be relaxed in the future to allow users to specify their own functions. However, this
+change will require a more thorough security review.
+
+# Drawbacks
+
+Although a "normal use" of PromptNodes would not trigger a need for InputOutputShaper there are cases where its
+use is necessary. In cases where we can only use pipeline definitions to configure the pipeline, we need to use it.
+
+
+- Implementation cost is not high. The InputOutputShaper component is already implemented and tested.
+- InputOutputShaper can't be defined outside of Haystack.
+- InputOutputShaper could turn out to be useful in other use cases as well - i.e. declarative pre/post processing.
+- InputOutputShaper integrates well with PromptNodes and other components.
+- No braking changes to existing components.
+
+
+# Alternatives
+
+A better solution would likely be a more general run method for components. This would allow us to arbitrarily
+define the pipeline invocation context. However even in those cases we'll need to use InputOutputShaper to modify
+existing variables in invocation context as needed.
+
+# Adoption strategy
+
+Haystack users can start using InputOutputShaper in their pipelines immediately. There are no breaking changes to
+existing components or pipelines.
+
+
+# How we teach this
+
+We will need docs update to teach users how to use InputOutputShaper. The docs will need to explain the
+motivation using InputOutputShaper and PromptNode examples. We also need to show the usage via tutorials.
+
+
+# Unresolved questions
+
+Optional, but suggested for first drafts. What parts of the design are still
+TBD?
```

Comment:
Let's not forget about the `run_batch` methods of the new node. I don't foresee any complications with that but so far it's not mentioned at all.